File: src/main/java/frc/robot/subsystems/Climber.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import edu.wpi.first.wpilibj.DigitalInput;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

import com.revrobotics.CANSparkMax;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.CANSparkMax.IdleMode;
import com.revrobotics.CANSparkMaxLowLevel.MotorType;

import frc.robot.Constants;

public class Climber extends SubsystemBase {
    private CANSparkMax climberMotor;
    private RelativeEncoder climberEncoder;
    private DigitalInput climberZeroSwitch;
    private int smartdashboardCounter = 0;

    /**
     * Creates the Climber subsystem with a single SparkMax NEO motor and a DIO
     * magnetic limit switch.
     */
    public Climber() {
        climberMotor = new CANSparkMax(Constants.Ports.CLIMBER_PORT, MotorType.kBrushless);
        climberMotor.restoreFactoryDefaults();
        climberMotor.setIdleMode(IdleMode.kBrake);
        climberMotor.enableVoltageCompensation(Constants.MAX_VOLTAGE_DRAW_VOLTS);

        climberEncoder = climberMotor.getEncoder();

        climberZeroSwitch = new DigitalInput(Constants.Ports.ZERO_SWITCH_PORT);
    }

    public double getMotorCurrent() {
        return climberMotor.getOutputCurrent();
    }

    public boolean getZeroSwitch() {
        return !climberZeroSwitch.get();
    }

    /**
     * Gets the climber encoder's current position. While it is not an absolute
     * encoder, the climber starts the match and plays most of teleop at or very
     * close to its zero position and the magnetic switch that zeros the encoder,
     * meaning that there will not be much offset.
     * 
     * @return The relative position of the encoder, in rotations.
     */
    public double getEncoderPosition() {
        return climberEncoder.getPosition();
    }

    /**
     * Sets the speed of the climber motor.
     * 
     * @param movement : The value between [-1,1] (inclusive) to send to the motor
     *                 controller.
     * @param override : Whether to ignore soft-stop protections.
     */
    public void climberMove(double movement, boolean override) {
        if (getZeroSwitch()) {
            resetEncoder();
        }
        if ((Math.abs(movement) <= Constants.Controller.DEAD_BAND) ||
                (!override && climberEncoder.getPosition() >= Constants.Climber.CLIMBER_MAX_HEIGHT && movement < 0) ||
                (!override && getZeroSwitch() && movement > 0)) {
            movement = 0;
        }

        climberMotor.set(movement);
    }

    /**
     * Sets the encoder of the climber motor to zero.
     */
    public void resetEncoder() {
        climberEncoder.setPosition(0);
    }

    @Override
    public void periodic() {
        // the counter is in place to prevent constant sending of the zero switch's
        // status over NetworkTables, using some amount of the bandwith the USB camera
        // would like to have. Other SmartDashboard calls that don't have this feature
        // are in place because they are for debug, not actual match play.
        if (smartdashboardCounter >= 10) {
            SmartDashboard.putBoolean("Zero Switch", getZeroSwitch());
            smartdashboardCounter = 0;
        }
        smartdashboardCounter += 1;
    }
}

File: src/main/cpp/subsystems/Climber.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <frc/smartdashboard/SmartDashboard.h>
#include <rev/CANSparkMax.h>
#include <algorithm>
#include "subsystems/Climber.h"
#include "Constants.h"

Climber::Climber() {
    m_climberMotor = new rev::CANSparkMax(constants::Ports::CLIMBER, rev::CANSparkMaxLowLevel::MotorType::kBrushless);
    m_climberMotor -> SetIdleMode(rev::CANSparkMax::IdleMode::kBrake);
    m_sensor = new frc::DigitalInput(constants::Ports::CLIMBER_SENSOR);
    m_encoder = new rev::SparkMaxRelativeEncoder(m_climberMotor -> GetEncoder());
    isSet = false;
}

void Climber::Set(double input, bool override) {
    input = -std::clamp(input, -constants::Values::CLIMBER_MAX_SPEED, constants::Values::CLIMBER_MAX_SPEED);
    frc::SmartDashboard::PutBoolean("CLIMBER SENSOR", !(m_sensor -> Get()));
    if (!(m_sensor -> Get())) {
        maximumPosition = -(m_encoder -> GetPosition()) + constants::Values::CLIMBER_UPPER_LIMIT;
        isSet = true;
    }
    if (!override && ((abs(input) <= constants::Values::CLIMBER_INPUT_DEADZONE) || 
    (((-(m_encoder -> GetPosition()) - maximumPosition < 0 || !isSet) && input <= 0)) ||
    (!(m_sensor -> Get()) && input >= 0))) {
        input = 0;
    } 
    frc::SmartDashboard::PutNumber("climber encoder value", m_encoder -> GetPosition());
    frc::SmartDashboard::PutNumber("climber max pos", maximumPosition);
    m_climberMotor -> Set(input);
}

Climber::~Climber() {
    delete m_climberMotor;
    delete m_encoder;
}

// This method will be called once per scheduler run
void Climber::Periodic() {}
